addons/stock/models/product_strategy.py

```diff
@@ -13,6 +13,18 @@ class RemovalStrategy(models.Model):
     name = fields.Char('Name', required=True, translate=True)
     method = fields.Char("Method", required=True, translate=True, help="FIFO, LIFO...")
 
+class ProductLastUsedPlacement(models.Model):
+    _name = 'product.last_used_placement'
+    _description = 'Stores the last location used for a product in a last used putaway strategy'
+
+    putaway_rule_id = fields.Many2one('stock.putaway.rule', check_company=True, required=True, ondelete='cascade')
+    product_id = fields.Many2one(
+        'product.product', 'Product', check_company=True, required=True,
+        domain="[('product_tmpl_id', '=', context.get('active_id', False))] if context.get('active_model') == 'product.template' else [('type', '!=', 'service')]",
+        ondelete='cascade')
+    last_location_id = fields.Many2one(
+        'stock.location', 'Last used location of this product in this putaway rule',
+        check_company=True, required=True, ondelete='cascade')
```

Comment:
Seems too complex to me. I would simply search the last stock move line to get the stock location once the putaway with the `last used` setting is triggered